haiopy/buffers.py

```diff
@@ -0,0 +1,218 @@
+        """Set the block size in samples. Only integer values are supported"""
+        self._set_block_size(block_size)
+
+    @abstractproperty
+    def sampling_rate(self):
+        """Return sampling rate."""
+        pass
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """Next dunder method for iteration"""
+        self._start()
+        return self.next()
+
+    @abstractmethod
+    def next(self):
+        """Next method which for sub-class specific handling of data."""
+        raise NotImplementedError()
+
+    @property
+    def is_active(self):
+        """Return the state of the buffer.
+        `True` if the buffer is active, `False` if inactive."""
+        return self._is_active
+
+    def check_if_active(self):
+        """Check if the buffer is active.
+        If the buffer is active a BufferError exception is raised. In case the
+        buffer is currently inactive, the method simply passes without any
+        return value.
+
+        Raises
+        ------
+        BufferError
+            Exception is raised if the buffer is currently active.
+        """
+        if self.is_active:
+            raise BufferError(
+                "The buffer needs to be inactive to be modified.")
+
+    def _stop(self, msg="Buffer iteration stopped."):
+        """Stop buffer iteration and set the state to inactive."""
+        self._is_active = False
+        raise StopIteration(msg)
+
+    def _start(self):
+        """Set the state to active.
+        Additional operations required before iterating the sub-class can be
+        implemented in the respective sub-class."""
+        self._is_active = True
+
+    def _reset(self):
+        """Stop and reset the buffer.
+        Resetting the buffer is implemented in the respective sub-class"""
+        self._stop()
+
+
+class SignalBuffer(Buffer):
+    """Buffer to block wise iterate a `pyfar.Signal`
+
+    Examples
+    --------
+
+    >>> import pyfar as pf
+    >>> from haiopy.buffers import SignalBuffer
+    >>> block_size = 512
+    >>> sine = pf.signals.sine(440, 4*block_size)
+    >>> buffer = SignalBuffer(blockk_size, sine)
+    >>> for block in buffer:
+    >>>     print(block)
+
+
+    """
+
+    def __init__(self, block_size, signal) -> None:
+        """Initialize a `SignalBuffer` with a given block size from a
+        `pyfar.Signal`.
+        If the number of audio samples is not an integer multiple of the
+        block size, the last block will be filled with zeros.
+
+        Parameters
+        ----------
+        block_size : int
+            The block size in samples
+        signal : pyfar.Signal
+            The audio data to be block wise iterated.
+
+        """
+        super().__init__(block_size)
+        if not isinstance(signal, pf.Signal):
+            raise ValueError("signal must be a pyfar.Signal object.")
+        if signal.time.ndim > 2:
+            raise ValueError("Only two-dimensional arrays are allowed")
```

Comment:
Maybe refer to the `cshape` here?
```suggestion
            raise ValueError("Only one-dimensional signals are allowed")
```

Reply:
True

Reply:
I'd suggest to refer to signals here instead of arrays